File: python/string_compression.py
"""
Implement a method to perform basic string compression using the counts of
repeated characters.
"""

# time complexity: O(n), space complexity: O(n)
def string_compression(string):
    if len(string) <= 2:
        return string

    compressed_strings = []
    repetitions_count = 0

    for i in range(len(string)):
        repetitions_count += 1
        if i == len(string) - 1 or string[i] != string[i + 1]:
            compressed_strings.append("%s%d" % (string[i], repetitions_count))
            repetitions_count = 0

    result = "".join(compressed_strings)
    return result if len(result) < len(string) else string


print(string_compression("aabcccccaaa"))

File: java/StringCompression.java
/*
Implement a method to perform basic string compression using the counts of
repeated characters.
*/

class StringCompression {
    public static void main(String args[]) {
        System.out.println(compressString("aabcccccaaa"));
    }

    // time complexity: O(n), space complexity: O(n)
    private static String compressString(String str) {
        if (str.length() <= 2) {
            return str;
        }
        StringBuilder resultBuilder = new StringBuilder();
        int charRepetitions = 0;

        for (int i = 0; i < str.length(); i++) {
            charRepetitions += 1;
            if (i == str.length() - 1 || str.charAt(i) != str.charAt(i + 1)) {
                resultBuilder.append(String.format("%c%d", str.charAt(i), charRepetitions));
                charRepetitions = 0;
            }
        }
        String result = resultBuilder.toString();
        return str.length() <= result.length() ? str : result;
    }
}
